File: frontend/src/types/prescription.ts
// —— Prescription Types ——————————————————————————————————————————

export type PrescriptionStatus = 'draft' | 'finalized' | 'partially_dispensed' | 'dispensed';
export type DurationUnit = 'days' | 'weeks' | 'months';
export type MedicineRoute = 'oral' | 'topical' | 'injection' | 'inhalation' | 'sublingual' | 'rectal' | 'nasal' | 'ophthalmic' | 'otic';
export type MedicineCategory = 'tablet' | 'capsule' | 'syrup' | 'injection' | 'cream' | 'drops' | 'ointment' | 'inhaler' | 'powder' | 'suspension';
/** Eye Hospital Drug Prescription format — which eye a medicine applies to. */
export type EyeSide = 'RE' | 'LE' | 'Both';

// —— Prescription Item ——————————————————————————————————————————

export interface PrescriptionItem {
  id: string;
  prescription_id: string;
  medicine_id: string | null;
  medicine_name: string;
  generic_name: string | null;
  dosage: string;
  frequency: string;
  duration_value: number | null;
  duration_unit: DurationUnit | null;
  route: MedicineRoute | null;
  instructions: string | null;
  quantity: number | null;
  allow_substitution: boolean;
  is_dispensed: boolean;
  dispensed_quantity: number;
  display_order: number;
  eye_side?: EyeSide | null;
  created_at: string;
}

export interface PrescriptionItemCreate {
  medicine_id?: string | null;
  medicine_name: string;
  generic_name?: string | null;
  dosage: string;
  frequency: string;
  duration_value?: number | null;
  duration_unit?: DurationUnit | null;
  route?: MedicineRoute | null;
  instructions?: string | null;
  quantity?: number | null;
  allow_substitution?: boolean;
  display_order?: number;
  eye_side?: EyeSide | null;
}

// —— Prescription ———————————————————————————————————————————————

export interface Prescription {
  id: string;
  hospital_id: string;
  prescription_number: string;
  appointment_id: string | null;
  patient_id: string;
  doctor_id: string | null;
  diagnosis: string | null;
  clinical_notes: string | null;
  advice: string | null;
  vitals_bp: string | null;
  vitals_pulse: string | null;
  vitals_temp: string | null;
  vitals_weight: string | null;
  vitals_spo2: string | null;
  vitals_blood_sugar: string | null;
  follow_up_date: string | null;
  queue_id: string | null;
  institution_id: string | null;
  institution_name?: string | null;
  is_opthal: boolean | null;
  opthal_notes: string | null;
  version: number;
  status: PrescriptionStatus;
  is_finalized: boolean;
  finalized_at: string | null;
  valid_until: string | null;
  created_by: string | null;
  is_deleted: boolean;
  created_at: string;
  updated_at: string;
  // Enriched fields
  patient_name?: string | null;
  patient_reference_number?: string | null;
  patient_gender?: string | null;
  patient_date_of_birth?: string | null;
  patient_age?: number | null;
  patient_blood_group?: string | null;
  patient_phone?: string | null;
  patient_email?: string | null;
  patient_known_allergies?: string | null;
  patient_chronic_conditions?: string | null;
  appointment_number?: string | null;
  doctor_name?: string | null;
  consultation_invoice_id?: string | null;
  consultation_invoice_number?: string | null;
  consultation_invoice_status?: string | null;
  final_amount?: number | null;
  dispensed_at?: string | null;
  items: PrescriptionItem[];
}

export interface PrescriptionListItem {
  id: string;
  prescription_number: string;
  patient_id: string;
  doctor_id: string | null;
  appointment_id?: string | null;
  diagnosis: string | null;
  status: PrescriptionStatus;
  is_finalized: boolean;
  is_opthal?: boolean | null;
  item_count: number;
  created_at: string;
  updated_at: string;
  patient_name?: string | null;
  patient_reference_number?: string | null;
  appointment_number?: string | null;
  doctor_name?: string | null;
  // Prescription Dashboard (BRD_OP_1 §3.1)
  chief_complaint?: string | null;
  medicine_names?: string[];
  dispensed_medicine_names?: string[];
  billed_qty?: number;
  billed_cost?: number;
  is_email_verified?: boolean;
  is_phone_verified?: boolean;
  is_verified?: boolean;
}

export interface PrescriptionCreate {
  patient_id: string;
  doctor_id?: string | null;
  appointment_id?: string | null;
  diagnosis?: string | null;
  clinical_notes?: string | null;
  advice?: string | null;
  vitals_bp?: string | null;
  vitals_pulse?: string | null;
  vitals_temp?: string | null;
  vitals_weight?: string | null;
  vitals_spo2?: string | null;
  vitals_blood_sugar?: string | null;
  follow_up_date?: string | null;
  queue_id?: string | null;
  valid_until?: string | null;
  institution_id?: string | null;
  is_opthal?: boolean | null;
  opthal_notes?: string | null;
  items: PrescriptionItemCreate[];
}

export interface PrescriptionUpdate {
  diagnosis?: string | null;
  clinical_notes?: string | null;
  advice?: string | null;
  vitals_bp?: string | null;
  vitals_pulse?: string | null;
  vitals_temp?: string | null;
  vitals_weight?: string | null;
  vitals_spo2?: string | null;
  vitals_blood_sugar?: string | null;
  follow_up_date?: string | null;
  valid_until?: string | null;
  institution_id?: string | null;
  is_opthal?: boolean | null;
  opthal_notes?: string | null;
  items?: PrescriptionItemCreate[] | null;
}

// Narrow, vitals-only write — backs PUT /prescriptions/draft-vitals, used by
// the nurse's Vitals entry screen (NurseVitals.tsx). No diagnosis/items/
// finalize field exists here on purpose — see PrescriptionVitalsUpdate on
// the backend.
export interface PrescriptionVitalsUpdate {
  patient_id: string;
  appointment_id: string;
  vitals_bp?: string;
  vitals_pulse?: string;
  vitals_temp?: string;
  vitals_weight?: string;
  vitals_spo2?: string;
  vitals_blood_sugar?: string;
}

// —— Medicine ———————————————————————————————————————————————————

export interface Medicine {
  id: string;
  hospital_id: string;
  name: string;
  generic_name: string;
  category: MedicineCategory | null;
  manufacturer: string | null;
  composition: string | null;
  strength: string | null;
  unit_of_measure: string;
  units_per_pack: number;
  requires_prescription: boolean;
  is_controlled: boolean;
  selling_price: number;
  purchase_price: number | null;
  reorder_level: number;
  total_stock?: number | null;
  storage_instructions: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface MedicineCreate {
  name: string;
  generic_name: string;
  category?: MedicineCategory | null;
  manufacturer?: string | null;
  composition?: string | null;
  strength?: string | null;
  unit_of_measure?: string;
  units_per_pack?: number;
  requires_prescription?: boolean;
  is_controlled?: boolean;
  selling_price: number;
  purchase_price?: number | null;
  reorder_level?: number;
  storage_instructions?: string | null;
}

// —— Prescription Template ——————————————————————————————————————

export interface PrescriptionTemplate {
  id: string;
  doctor_id: string;
  name: string;
  diagnosis: string | null;
  items: TemplateItem[];
  advice: string | null;
  is_active: boolean;
  usage_count: number;
  created_at: string;
  updated_at: string;
}

export interface TemplateItem {
  medicine_name: string;
  generic_name?: string | null;
  dosage: string;
  frequency: string;
  duration_value?: number | null;
  duration_unit?: DurationUnit | null;
  route?: MedicineRoute | null;
  instructions?: string | null;
}

export interface PrescriptionTemplateCreate {
  name: string;
  diagnosis?: string | null;
  items: TemplateItem[];
  advice?: string | null;
}

// —— Prescription Version ——————————————————————————————————————

export interface PrescriptionVersion {
  id: string;
  prescription_id: string;
  version: number;
  snapshot: Record<string, unknown>;
  changed_by: string | null;
  change_reason: string | null;
  created_at: string;
}

// —— Generic Paginated (reuse pattern) ————————————————————————

export interface PaginatedResponse<T> {
  total: number;
  page: number;
  limit: number;
  total_pages: number;
  data: T[];
}

File: backend/app/schemas/prescription.py
"""
Pydantic schemas for the Prescription module.
Follows the same patterns as appointment.py schemas.
"""
from pydantic import BaseModel, Field, field_validator, model_validator, ConfigDict
from typing import Optional, Any
from datetime import date, datetime


VALID_PRESCRIPTION_STATUSES = ["draft", "finalized", "dispensed", "partially_dispensed"]
VALID_PRESCRIPTION_TYPES = ["general", "optical"]
VALID_DURATION_UNITS = ["days", "weeks", "months"]
VALID_ROUTES = ["oral", "topical", "injection", "inhalation", "sublingual", "rectal", "nasal", "ophthalmic", "otic"]
VALID_LENS_TYPES = ["single_vision", "bifocal", "progressive", "contact"]
VALID_MEDICINE_CATEGORIES = ["tablet", "capsule", "syrup", "injection", "cream", "drops", "ointment", "inhaler", "powder", "suspension"]


# —— Helper: ORM -> dict with stringified UUIDs ——

def _orm_to_dict(data: Any) -> Any:
    """Convert SQLAlchemy model instance to dict with all UUIDs as strings."""
    if hasattr(data, "__table__"):
        d = {}
        for col in data.__table__.columns:
            val = getattr(data, col.name)
            if hasattr(val, "hex"):  # uuid.UUID
                val = str(val)
            d[col.name] = val
        return d
    if isinstance(data, dict):
        return data
    return data


# ═══════════════════════════════════════════════════════════════════════════
# Medicine Schemas
# ═══════════════════════════════════════════════════════════════════════════

class MedicineCreate(BaseModel):
    name: str = Field(..., max_length=200)
    generic_name: str = Field(..., max_length=200)
    category: Optional[str] = None
    manufacturer: Optional[str] = None
    composition: Optional[str] = None
    strength: Optional[str] = None
    unit_of_measure: str = Field(default="strip", max_length=20)
    units_per_pack: int = Field(default=1, ge=1)
    requires_prescription: bool = True
    is_controlled: bool = False
    selling_price: float = Field(..., ge=0)
    purchase_price: Optional[float] = None
    reorder_level: int = Field(default=10, ge=0)
    storage_instructions: Optional[str] = None

    @field_validator("category")
    @classmethod
    def validate_category(cls, v: Optional[str]) -> Optional[str]:
        if v and v not in VALID_MEDICINE_CATEGORIES:
            raise ValueError(f"Must be one of: {', '.join(VALID_MEDICINE_CATEGORIES)}")
        return v


class MedicineUpdate(BaseModel):
    name: Optional[str] = None
    generic_name: Optional[str] = None
    category: Optional[str] = None
    manufacturer: Optional[str] = None
    composition: Optional[str] = None
    strength: Optional[str] = None
    unit_of_measure: Optional[str] = None
    selling_price: Optional[float] = None
    purchase_price: Optional[float] = None
    reorder_level: Optional[int] = None
    is_active: Optional[bool] = None
    storage_instructions: Optional[str] = None


class MedicineResponse(BaseModel):
    id: str
    hospital_id: str
    name: str
    generic_name: str
    category: Optional[str] = None
    manufacturer: Optional[str] = None
    composition: Optional[str] = None
    strength: Optional[str] = None
    unit_of_measure: str
    units_per_pack: int = 1
    requires_prescription: bool = True
    is_controlled: bool = False
    selling_price: float
    purchase_price: Optional[float] = None
    reorder_level: int = 10
    storage_instructions: Optional[str] = None
    is_active: bool = True
    created_at: datetime
    updated_at: datetime

    @model_validator(mode="before")
    @classmethod
    def transform(cls, data: Any) -> Any:
        return _orm_to_dict(data)

    model_config = ConfigDict(from_attributes=True)


class PaginatedMedicineResponse(BaseModel):
    total: int
    page: int
    limit: int
    total_pages: int
    data: list[MedicineResponse]


# ═══════════════════════════════════════════════════════════════════════════
# Prescription Item Schemas
# ═══════════════════════════════════════════════════════════════════════════

class PrescriptionItemCreate(BaseModel):
    medicine_id: Optional[str] = None
    medicine_name: str = Field(..., max_length=200)
    generic_name: Optional[str] = None
    dosage: str = Field(..., max_length=50)
    frequency: str = Field(..., max_length=50)  # e.g., '1-0-1'
    duration_value: Optional[int] = Field(None, ge=1)
    duration_unit: Optional[str] = None
    route: Optional[str] = None
    instructions: Optional[str] = None
    quantity: Optional[int] = Field(None, ge=1)
    allow_substitution: bool = True
    display_order: int = 0

    @field_validator("duration_unit")
    @classmethod
    def validate_duration_unit(cls, v: Optional[str]) -> Optional[str]:
        if v and v not in VALID_DURATION_UNITS:
            raise ValueError(f"Must be one of: {', '.join(VALID_DURATION_UNITS)}")
        return v

    @field_validator("route")
    @classmethod
    def validate_route(cls, v: Optional[str]) -> Optional[str]:
        if v and v not in VALID_ROUTES:
            raise ValueError(f"Must be one of: {', '.join(VALID_ROUTES)}")
        return v


class PrescriptionItemUpdate(BaseModel):
    medicine_name: Optional[str] = None
    generic_name: Optional[str] = None
    dosage: Optional[str] = None
    frequency: Optional[str] = None
    duration_value: Optional[int] = None
    duration_unit: Optional[str] = None
    route: Optional[str] = None
    instructions: Optional[str] = None
    quantity: Optional[int] = None
    allow_substitution: Optional[bool] = None
    display_order: Optional[int] = None


class PrescriptionItemResponse(BaseModel):
    id: str
    prescription_id: str
    medicine_id: Optional[str] = None
    medicine_name: str
    generic_name: Optional[str] = None
    dosage: str
    frequency: str
    duration_value: Optional[int] = None
    duration_unit: Optional[str] = None
    route: Optional[str] = None
    instructions: Optional[str] = None
    quantity: Optional[int] = None
    allow_substitution: bool = True
    is_dispensed: bool = False
    dispensed_quantity: int = 0
    display_order: int = 0
    created_at: datetime

    @model_validator(mode="before")
    @classmethod
    def transform(cls, data: Any) -> Any:
        return _orm_to_dict(data)

    model_config = ConfigDict(from_attributes=True)


# ═══════════════════════════════════════════════════════════════════════════
# Prescription Schemas
# ═══════════════════════════════════════════════════════════════════════════

class PrescriptionCreate(BaseModel):
    patient_id: str
    doctor_id: Optional[str] = None  # Defaults to current doctor
    appointment_id: Optional[str] = None
    prescription_type: str = Field(default="general", max_length=20)
    diagnosis: Optional[str] = None
    clinical_notes: Optional[str] = None
    advice: Optional[str] = None
    vitals_bp: Optional[str] = None
    vitals_pulse: Optional[str] = None
    vitals_temp: Optional[str] = None
    # Optical fields
    right_sphere: Optional[str] = None
    right_cylinder: Optional[str] = None
    right_axis: Optional[str] = None
    right_add: Optional[str] = None
    right_va: Optional[str] = None
    right_ipd: Optional[str] = None
    left_sphere: Optional[str] = None
    left_cylinder: Optional[str] = None
    left_axis: Optional[str] = None
    left_add: Optional[str] = None
    left_va: Optional[str] = None
    left_ipd: Optional[str] = None
    lens_type: Optional[str] = None
    lens_material: Optional[str] = None
    lens_coating: Optional[str] = None
    optical_notes: Optional[str] = None

    @field_validator("prescription_type")
    @classmethod
    def validate_prescription_type(cls, v: str) -> str:
        if v not in VALID_PRESCRIPTION_TYPES:
            raise ValueError(f"Must be one of: {', '.join(VALID_PRESCRIPTION_TYPES)}")
        return v

    @field_validator("lens_type")
    @classmethod
    def validate_lens_type(cls, v: Optional[str]) -> Optional[str]:
        if v and v not in VALID_LENS_TYPES:
            raise ValueError(f"Must be one of: {', '.join(VALID_LENS_TYPES)}")
        return v
    vitals_weight: Optional[str] = None
    vitals_spo2: Optional[str] = None
    follow_up_date: Optional[date] = None
    queue_id: Optional[str] = None
    valid_until: Optional[date] = None
    items: list[PrescriptionItemCreate] = Field(default_factory=list)


class PrescriptionUpdate(BaseModel):
    diagnosis: Optional[str] = None
    clinical_notes: Optional[str] = None
    advice: Optional[str] = None
    vitals_bp: Optional[str] = None
    vitals_pulse: Optional[str] = None
    vitals_temp: Optional[str] = None
    vitals_weight: Optional[str] = None
    vitals_spo2: Optional[str] = None
    follow_up_date: Optional[date] = None
    valid_until: Optional[date] = None
    items: Optional[list[PrescriptionItemCreate]] = None
    # Optical fields
    right_sphere: Optional[str] = None
    right_cylinder: Optional[str] = None
    right_axis: Optional[str] = None
    right_add: Optional[str] = None
    right_va: Optional[str] = None
    right_ipd: Optional[str] = None
    left_sphere: Optional[str] = None
    left_cylinder: Optional[str] = None
    left_axis: Optional[str] = None
    left_add: Optional[str] = None
    left_va: Optional[str] = None
    left_ipd: Optional[str] = None
    lens_type: Optional[str] = None
    lens_material: Optional[str] = None
    lens_coating: Optional[str] = None
    optical_notes: Optional[str] = None


class PrescriptionResponse(BaseModel):
    id: str
    hospital_id: str
    prescription_number: str
    prescription_type: str = "general"
    appointment_id: Optional[str] = None
    patient_id: str
    doctor_id: str
    diagnosis: Optional[str] = None
    clinical_notes: Optional[str] = None
    advice: Optional[str] = None
    vitals_bp: Optional[str] = None
    vitals_pulse: Optional[str] = None
    vitals_temp: Optional[str] = None
    vitals_weight: Optional[str] = None
    vitals_spo2: Optional[str] = None
    follow_up_date: Optional[date] = None
    # Optical fields
    right_sphere: Optional[str] = None
    right_cylinder: Optional[str] = None
    right_axis: Optional[str] = None
    right_add: Optional[str] = None
    right_va: Optional[str] = None
    right_ipd: Optional[str] = None
    left_sphere: Optional[str] = None
    left_cylinder: Optional[str] = None
    left_axis: Optional[str] = None
    left_add: Optional[str] = None
    left_va: Optional[str] = None
    left_ipd: Optional[str] = None
    lens_type: Optional[str] = None
    lens_material: Optional[str] = None
    lens_coating: Optional[str] = None
    optical_notes: Optional[str] = None
    queue_id: Optional[str] = None
    version: int = 1
    status: str = "draft"
    is_finalized: bool = False
    finalized_at: Optional[datetime] = None
    valid_until: Optional[date] = None
    created_by: Optional[str] = None
    is_deleted: bool = False
    created_at: datetime
    updated_at: datetime
    # Enriched fields
    patient_name: Optional[str] = None
    patient_reference_number: Optional[str] = None
    appointment_number: Optional[str] = None
    doctor_name: Optional[str] = None
    items: list[PrescriptionItemResponse] = []

    @model_validator(mode="before")
    @classmethod
    def transform(cls, data: Any) -> Any:
        if isinstance(data, dict):
            return data
        return _orm_to_dict(data)

    model_config = ConfigDict(from_attributes=True)


class PrescriptionListItem(BaseModel):
    id: str
    prescription_number: str
    prescription_type: str = "general"
    patient_id: str
    doctor_id: str
    diagnosis: Optional[str] = None
    status: str = "draft"
    is_finalized: bool = False
    item_count: int = 0
    created_at: datetime
    updated_at: datetime
    patient_name: Optional[str] = None
    patient_reference_number: Optional[str] = None
    appointment_number: Optional[str] = None
    doctor_name: Optional[str] = None

    @model_validator(mode="before")
    @classmethod
    def transform(cls, data: Any) -> Any:
        if isinstance(data, dict):
            return data
        return _orm_to_dict(data)

    model_config = ConfigDict(from_attributes=True)


class PaginatedPrescriptionResponse(BaseModel):
    total: int
    page: int
    limit: int
    total_pages: int
    data: list[PrescriptionListItem]


# ═══════════════════════════════════════════════════════════════════════════
# Prescription Template Schemas
# ═══════════════════════════════════════════════════════════════════════════

class TemplateItemSchema(BaseModel):
    medicine_name: str
    generic_name: Optional[str] = None
    dosage: str
    frequency: str
    duration_value: Optional[int] = None
    duration_unit: Optional[str] = None
    route: Optional[str] = None
    instructions: Optional[str] = None


class PrescriptionTemplateCreate(BaseModel):
    name: str = Field(..., max_length=100)
    diagnosis: Optional[str] = None
    items: list[TemplateItemSchema]
    advice: Optional[str] = None


class PrescriptionTemplateUpdate(BaseModel):
    name: Optional[str] = None
    diagnosis: Optional[str] = None
    items: Optional[list[TemplateItemSchema]] = None
    advice: Optional[str] = None
    is_active: Optional[bool] = None


class PrescriptionTemplateResponse(BaseModel):
    id: str
    doctor_id: str
    name: str
    diagnosis: Optional[str] = None
    items: list[dict]
    advice: Optional[str] = None
    is_active: bool = True
    usage_count: int = 0
    created_at: datetime
    updated_at: datetime

    @model_validator(mode="before")
    @classmethod
    def transform(cls, data: Any) -> Any:
        return _orm_to_dict(data)

    model_config = ConfigDict(from_attributes=True)


# ═══════════════════════════════════════════════════════════════════════════
# Prescription Version Schemas
# ═══════════════════════════════════════════════════════════════════════════

class PrescriptionVersionResponse(BaseModel):
    id: str
    prescription_id: str
    version: int
    snapshot: dict
    changed_by: Optional[str] = None
    change_reason: Optional[str] = None
    created_at: datetime

    @model_validator(mode="before")
    @classmethod
    def transform(cls, data: Any) -> Any:
        return _orm_to_dict(data)

    model_config = ConfigDict(from_attributes=True)
